modules/host/negotiaterenewcontract.go

```diff
@@ -0,0 +1,315 @@
+func (h *Host) managedAddRenewCollateral(so *storageObligation, settings modules.HostExternalSettings, txnSet []types.Transaction) (builder modules.TransactionBuilder, newParents []types.Transaction, newInputs []types.SiacoinInput, newOutputs []types.SiacoinOutput, err error) {
+	hostPortion := renewContractCollateral(so, settings, txnSet)
+	txn := txnSet[len(txnSet)-1]
+	parents := txnSet[:len(txnSet)-1]
+	builder = h.wallet.RegisterTransaction(txn, parents)
+	err = builder.FundSiacoins(hostPortion)
+	if err != nil {
+		builder.Drop()
+		return nil, nil, nil, nil, err
+	}
+
+	// Return which inputs and outputs have been added by the collateral call.
+	newParentIndices, newInputIndices, newOutputIndices, _ := builder.ViewAdded()
+	updatedTxn, updatedParents := builder.View()
+	for _, parentIndex := range newParentIndices {
+		newParents = append(newParents, updatedParents[parentIndex])
+	}
+	for _, inputIndex := range newInputIndices {
+		newInputs = append(newInputs, updatedTxn.SiacoinInputs[inputIndex])
+	}
+	for _, outputIndex := range newOutputIndices {
+		newOutputs = append(newOutputs, updatedTxn.SiacoinOutputs[outputIndex])
+	}
+	return builder, newParents, newInputs, newOutputs, nil
+}
+
+// managedRenewContract accepts a request to renew a file contract.
+func (h *Host) managedRPCRenewContract(conn net.Conn) error {
+	// Perform the recent revision protocol to get the file contract being
+	// revised.
+	_, so, err := h.managedRPCRecentRevision(conn)
+	if err != nil {
+		return err
+	}
+
+	// Lock the storage obligation for the remainder of the connection.
+	h.mu.Lock()
+	err = h.lockStorageObligation(so)
+	h.mu.Unlock()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		h.mu.Lock()
+		err = h.unlockStorageObligation(so)
+		h.mu.Unlock()
+		if err != nil {
+			h.log.Critical(err)
+		}
+	}()
+
+	// Perform the host settings exchange with the renter.
+	err = h.managedRPCSettings(conn)
+	if err != nil {
+		return err
+	}
+
+	// Set the renewal deadline.
+	conn.SetDeadline(time.Now().Add(modules.NegotiateRenewContractTime))
+
+	// The renter will either accept or reject the host's settings.
+	err = modules.ReadNegotiationAcceptance(conn)
+	if err != nil {
+		return err
+	}
+	// If the renter sends an acceptance of the settings, it will be followed
+	// by an unsigned transaction containing funding from the renter and a file
+	// contract which matches what the final file contract should look like.
+	// After the file contract, the renter will send a public key which is the
+	// renter's public key in the unlock conditions that protect the file
+	// contract from revision.
+	var txnSet []types.Transaction
+	var renterPK crypto.PublicKey
+	err = encoding.ReadObject(conn, &txnSet, modules.NegotiateMaxFileContractSetLen)
+	if err != nil {
+		return err
+	}
+	err = encoding.ReadObject(conn, &renterPK, modules.NegotiateMaxSiaPubkeySize)
+	if err != nil {
+		return err
+	}
+
+	h.mu.RLock()
+	settings := h.externalSettings()
+	h.mu.RUnlock()
```

Comment:
I thought about sending the settings around starting from the beginning of the connection, but this is a rare race condition and it will be handled mostly gracefully. The host will either reject the renter's file contract, or nothing unusual will happen. Host will only reject the renter if the price goes up, nothing happens if the price goes down.

This is something that can reasonably be addressed later. Might be a larger issue once the host is automatically adjusting prices to meet supply + demand.